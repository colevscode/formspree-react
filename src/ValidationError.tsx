import React from 'react';

interface Props {
  prefix?: string;
  field: string;
  errors: Array<{
    field: string;
    message: string;
    code: string | null;
    properties: object;
  }>;
  [x: string]: any;
}

export const ValidationError: React.FC<Props> = props => {
  const { prefix, field, errors, ...attrs } = props;

  const error = (errors || []).find(error => {
    return error.field == field;
  });

  if (!error) {
    return null;
  }

  return (
    <div {...attrs}>
      {prefix} {error.message}
    </div>
  );
};
