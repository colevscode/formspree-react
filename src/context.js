import React, { useEffect, useState, useContext } from 'react';
import createClient from '@statickit/core';

const StaticKitContext = React.createContext({
  client: null
});

StaticKitContext.displayName = 'StaticKit';

export function StaticKit(props) {
  const [client, setClient] = useState(props.client);
  const [site, _] = useState(props.site);

  if (!props.client && !props.site) {
    throw new Error('site prop is required');
  }

  useEffect(() => {
    if (!client) {
      setClient(createClient({ site }));
    }

    return () => {
      if (client) client.teardown();
    };
  }, [site]);

  return (
    <StaticKitContext.Provider value={{ client }}>
      {props.children}
    </StaticKitContext.Provider>
  );
}

export function useStaticKit() {
  const { client } = useContext(StaticKitContext);
  return client;
}
