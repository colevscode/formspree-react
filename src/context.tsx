import React, { useEffect, useState, useContext } from 'react';
import { StaticKit as Client, createClient } from '@statickit/core';

interface Context {
  client: undefined | Client;
}

export interface Props {
  site?: string;
  client?: Client;
}

const StaticKitContext = React.createContext<Context>({
  client: undefined
});

StaticKitContext.displayName = 'StaticKit';

export const StaticKit: React.FC<Props> = props => {
  const [client, setClient] = useState(props.client);
  const [site, _] = useState(props.site);

  if (!props.client && !props.site) {
    throw new Error('site prop is required');
  }

  useEffect(() => {
    if (!client) {
      setClient(createClient({ site }));
    }

    return () => {
      if (client) client.teardown();
    };
  }, [site]);

  return (
    <StaticKitContext.Provider value={{ client }}>
      {props.children}
    </StaticKitContext.Provider>
  );
};

export function useStaticKit(): Client | undefined {
  const { client } = useContext(StaticKitContext);
  return client;
}
