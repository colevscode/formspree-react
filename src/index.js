import React, { useState, useEffect } from 'react';
import StaticKit from '@statickit/core';

function ValidationError(props) {
  const { prefix, field, errors, ...attrs } = props;

  const error = (errors || []).find(error => {
    return error.field == field;
  });

  if (!error) {
    return null;
  }

  return (
    <div {...attrs}>
      {prefix} {error.message}
    </div>
  );
}

function useForm(props) {
  const [submitting, setSubmitting] = useState(false);
  const [succeeded, setSucceeded] = useState(false);
  const [errors, setErrors] = useState([]);
  const [client, setClient] = useState(undefined);

  useEffect(() => {
    const client = StaticKit();
    setClient(client);
  }, []);

  const id = typeof props === 'object' ? props.id : props;
  if (!id) throw new Error('You must define an `id` property');

  const endpoint = props.endpoint || 'https://api.statickit.com';
  const debug = !!props.debug;

  const submit = event => {
    const form = event.target;

    if (form.tagName != 'FORM')
      throw new Error('submit was triggered for a non-form element');

    const formData = new FormData(form);

    const sendRequest = async () => {
      try {
        const result = await client.submitForm({
          id: id,
          endpoint: endpoint,
          data: formData
        });

        if (result.response.status == 200) {
          if (debug) console.log(id, 'Submitted', result);
          setSucceeded(true);
          setErrors([]);
        } else {
          const errors = result.body.errors;
          if (debug) console.log(id, 'Validation error', result);
          setSucceeded(false);
          setErrors(errors);
        }
      } catch (e) {
        if (debug) console.log(id, 'Unexpected error', e);
        setSucceeded(false);
      } finally {
        setSubmitting(false);
      }
    };

    e.preventDefault();
    setSubmitting(true);
    sendRequest();
    return true;
  };

  return [{ submitting, succeeded, errors }, submit];
}

export { useForm, ValidationError };
