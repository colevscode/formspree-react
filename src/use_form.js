import { useState, useEffect, useRef } from 'react';
import StaticKit from '@statickit/core';

export default function useForm(props) {
  const [submitting, setSubmitting] = useState(false);
  const [succeeded, setSucceeded] = useState(false);
  const [errors, setErrors] = useState([]);
  const client = useRef(props.client);

  useEffect(() => {
    if (!client.current) {
      client.current = StaticKit();
    }

    return () => {
      if (client.current) {
        client.current.teardown();
      }
    };
  }, []);

  const id = typeof props === 'object' ? props.id : props;

  if (!id && !(props.site && props.form)) {
    throw new Error('You must set an `id` or `site` & `form` properties');
  }

  const endpoint = props.endpoint || 'https://api.statickit.com';
  const debug = !!props.debug;
  const extraData = props.data;

  const submit = event => {
    const form = event.target;

    if (form.tagName != 'FORM')
      throw new Error('submit was triggered for a non-form element');

    const formData = new FormData(form);

    // Append extra data from config
    if (typeof extraData === 'object') {
      for (const prop in extraData) {
        if (typeof extraData[prop] === 'function') {
          formData.append(prop, extraData[prop].call(null, config));
        } else {
          formData.append(prop, extraData[prop]);
        }
      }
    }

    event.preventDefault();
    setSubmitting(true);

    return client.current
      .submitForm({
        id: id,
        site: props.site,
        form: props.form,
        endpoint: endpoint,
        data: formData
      })
      .then(result => {
        switch (result.response.status) {
          case 200:
            if (debug) console.log('Form submitted', result);
            setSucceeded(true);
            setErrors([]);
            break;

          case 422:
            const errors = result.body.errors;
            if (debug) console.log('Validation error', result);
            setSucceeded(false);
            setErrors(errors);
            break;

          default:
            if (debug) console.log('Unexpected error', result);
            setSucceeded(false);
            break;
        }

        return result;
      })
      .catch(error => {
        if (debug) console.log(id, 'Unexpected error', error);
        setSucceeded(false);
      })
      .finally(() => {
        setSubmitting(false);
      });
  };

  return [{ submitting, succeeded, errors }, submit];
}
