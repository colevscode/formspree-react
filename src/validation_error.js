import React from 'react';

export default function ValidationError(props) {
  const { prefix, field, errors, ...attrs } = props;

  const error = (errors || []).find(error => {
    return error.field == field;
  });

  if (!error) {
    return null;
  }

  return (
    <div {...attrs}>
      {prefix} {error.message}
    </div>
  );
}
