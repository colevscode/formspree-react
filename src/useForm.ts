import React, { useState } from 'react';
import { useFormspree } from './context';
import { version } from '../package.json';
import { Client } from '@formspree/core';
import { SubmissionResponse } from '@formspree/core/forms';

type SubmitHandler = (
  event: React.FormEvent<HTMLFormElement>
) => Promise<SubmissionResponse>;

interface ErrorResponse {
  errors: Array<{
    field: string;
    message: string;
    code: string | null;
    properties: object;
  }>;
}

export function useForm(
  formKey: string,
  args: {
    client?: Client;
    data?: { [key: string]: string | (() => string) };
    endpoint?: string;
    debug?: boolean;
  } = {}
): [
  {
    submitting: boolean;
    succeeded: boolean;
    errors: {
      field: string;
      message: string;
      code: string | null;
      properties: object;
    }[];
  },
  SubmitHandler
] {
  const [submitting, setSubmitting] = useState(false);
  const [succeeded, setSucceeded] = useState(false);
  const [errors, setErrors] = useState([]);
  const globalClient = useFormspree();
  const client = args.client || globalClient;

  if (!client) {
    throw new Error('You must provide a Formspree client');
  }

  if (!formKey) {
    throw new Error('You must provide a form key (e.g. useForm("myForm")');
  }

  const debug = !!args.debug;
  const extraData = args.data;

  const handleSubmit: SubmitHandler = event => {
    event.preventDefault();
    const form = event.target as HTMLFormElement;

    if (form.tagName != 'FORM') {
      throw new Error('submit was triggered for a non-form element');
    }

    const formData = new FormData(form);

    // Append extra data from config
    if (typeof extraData === 'object') {
      for (const prop in extraData) {
        if (typeof extraData[prop] === 'function') {
          formData.append(prop, (extraData[prop] as (() => string)).call(null));
        } else {
          formData.append(prop, extraData[prop] as string);
        }
      }
    }

    setSubmitting(true);

    return client
      .submitForm(formKey, formData, {
        endpoint: args.endpoint,
        clientName: `@formspree/react@${version}`
      })
      .then((result: SubmissionResponse) => {
        switch (result.response.status) {
          case 200:
            if (debug) console.log('Form submitted', result);
            setSucceeded(true);
            setErrors([]);
            break;

          case 422:
            let body = result.body as ErrorResponse;
            let errors = body.errors;
            if (debug) console.log('Validation error', result);
            setSucceeded(false);
            setErrors(errors);
            break;

          default:
            if (debug) console.log('Unexpected error', result);
            setSucceeded(false);
            break;
        }

        return result;
      })
      .catch((error: Error) => {
        if (debug) console.log('Unexpected error', error);
        setSucceeded(false);
        throw error;
      })
      .finally(() => {
        setSubmitting(false);
      });
  };

  return [{ submitting, succeeded, errors }, handleSubmit];
}
