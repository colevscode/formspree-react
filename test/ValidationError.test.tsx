import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-dom/test-utils';
import { ValidationError } from '../src';

const { act } = ReactTestUtils;

let container: HTMLElement;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it('renders an error if one exists', () => {
  act(() => {
    ReactDOM.render(
      <ValidationError
        prefix="Email"
        field="email"
        className="error"
        errors={[
          {
            field: 'email',
            message: 'is required',
            code: 'REQUIRED',
            properties: {}
          }
        ]}
      />,
      container
    );
  });

  expect(container).toMatchSnapshot();
});

it('does not render anything if the field does not have an error', () => {
  act(() => {
    ReactDOM.render(
      <ValidationError
        prefix="Email"
        field="email"
        errors={[
          {
            field: 'name',
            message: 'is required',
            code: 'REQUIRED',
            properties: {}
          }
        ]}
      />,
      container
    );
  });

  expect(container).toMatchSnapshot();
});
