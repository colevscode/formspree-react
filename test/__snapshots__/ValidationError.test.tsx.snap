// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render anything if the field does not have an error 1`] = `<div />`;

exports[`renders an error if one exists 1`] = `
<div>
  <div
    class="error"
  >
    Email
     
    is required
  </div>
</div>
`;
